src/gravel/controllers/orch/nfs.py

```diff
@@ -78,23 +100,100 @@ def ls(self) -> List[str]:
+            daemons = parse_obj_as(List[NFSDaemonModel], res[service_id])
             ret.append(
                 NFSServiceModel(
-                    name=name,
+                    service_id=service_id,
                     daemons=daemons
                 )
             )
         return ret
+
+
+class NFSExport(NFS):
+    def create(
+            self,
+            service_id: str,
+            binding: str,
+            fs_type: NFSBackingStoreEnum,
+            fs_name: str,
+            fs_path: Optional[str],
+            readonly: bool = False) -> NFSExportModel:
+        binding = str(Path('/').joinpath(binding))
+        cmd = {
+            # TODO: fix upstream: prefix contains `fs_type`
+            # TODO: fix upstream: `rgw` is not currently supported
+            'prefix': f'nfs export create {fs_type}',
+            'fsname': fs_name,
+            'clusterid': service_id,
+            'binding': binding,
+            'readonly': readonly,
+        }
+        if fs_path:
+            cmd['path'] = fs_path
+
+        # TODO: fix upstream: add json formatter?
+        #       output is sometimes json, sometimes str
+        res = self._call(cmd)
+        if 'result' in res:
+            # TODO: fix upstream: return proper errno
+            raise NFSError(res['result'])
+
+        # find the newly created export
+        # TODO: fix upstream: return `export_id` in the create response?
+        for export in self.info(service_id):
+            # TODO: fix upstream: `binding`, `bind`, `path`, `pseudo` ??
+            if export.pseudo == binding:
+                return export
+
+        # cannot find the export, did an error occur?
+        if 'result' in res:
+            raise NFSError(res['result'])
+        raise NFSError(f'failed to create nfs export: {res}')
+
+    def delete(self, service_id: str, export_id: int):
+        for export in self.info(service_id):
+            if export.export_id == export_id:
+                res = self._call({
+                    'prefix': 'nfs export delete',
+                    'clusterid': service_id,
+                    # TODO: fix upstream: delete by `export_id`?
+                    'binding': str(Path('/').joinpath(export.pseudo)),
+                })
+                return res['result']
+        # TODO: return errno to create HTTP 404
+        raise NFSError('export not found')
+
+    def ls(self, service_id: str) -> List[int]:
+        return sorted([e.export_id for e in self.info(service_id)])
+
+    def info(self, service_id: str) -> List[NFSExportModel]:
+        cmd = {
+            'prefix': 'nfs export ls',
+            'clusterid': service_id,
+            'detailed': True,  # TODO: fix upstream: `detailed` -> `detail`?
+            'format': 'json',
+        }
+
+        res = self._call(cmd)
+
+        ret: List[NFSExportModel] = []
+        for export in res:
+            ret.append(
+                NFSExportModel(
+                    **export
```

Comment:
What happens if the data returned from `ceph nfs export ls` changes in future (extra fields, or whatever)? Does this still do the right thing? (Answer, having done a little further poking around: probably, yes, because pydantic should only pull in data that matches the model)

Reply:
yes, the extra args are ignored if they are not defined by the model. However, if the model defines an arg that does not exist, pydantic will raise an error.